contracts/integrations/nevermined/BalanceTrackerSubscription.sol

```diff
@@ -0,0 +1,200 @@
+error InsufficientBalance(uint256 current, uint256 required);
+
+/// @dev Caught reentrancy violation.
+error ReentrancyGuard();
+
+/// @dev Account is unauthorized.
+/// @param account Account address.
+error UnauthorizedAccount(address account);
+
+/// @dev No incoming msg.value is allowed.
+/// @param amount Value amount.
+error NoDepositAllowed(uint256 amount);
+
+contract BalanceTrackerSubscription is ERC1155TokenReceiver {
+    event MechPaymentCalculated(address indexed mech, uint256 indexed requestId, uint256 deliveryRate, uint256 rateDiff);
+    event CreditsAccounted(address indexed account, uint256 amount);
+
+    // Mech marketplace address
+    address public immutable mechMarketplace;
+    // Subscription NFT
+    address public immutable subscriptionNFT;
+    // Subscription token Id
+    uint256 public immutable subscriptionTokenId;
+
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    // Map of requester => current credit balance
+    mapping(address => uint256) public mapRequesterBalances;
+    // Map of mech => current debit balance
+    mapping(address => uint256) public mapMechBalances;
+
+    /// @dev BalanceTrackerSubscription constructor.
+    /// @param _mechMarketplace Mech marketplace address.
+    /// @param _subscriptionNFT Subscription NFT address.
+    /// @param _subscriptionTokenId Subscription token Id.
+    constructor(address _mechMarketplace, address _subscriptionNFT, uint256 _subscriptionTokenId) {
+        if (_subscriptionNFT == address(0)) {
+            revert ZeroAddress();
+        }
+
+        if (_subscriptionTokenId == 0) {
+            revert ZeroValue();
+        }
+
+        mechMarketplace = _mechMarketplace;
+        subscriptionNFT = _subscriptionNFT;
+        subscriptionTokenId = _subscriptionTokenId;
+    }
+
+    // Check and record delivery rate
+    function checkAndRecordDeliveryRate(
+        address mech,
+        address requester,
+        uint256,
+        bytes memory
+    ) external payable {
+        // Check for marketplace access
+        if (msg.sender != mechMarketplace) {
+            revert ManagerOnly(msg.sender, mechMarketplace);
+        }
+
+        // Get mech max delivery rate
+        uint256 maxDeliveryRate = IMech(mech).maxDeliveryRate();
+
+        // Check that there is no incoming deposit
+        if (msg.value > 0) {
+            revert NoDepositAllowed(msg.value);
+        }
+
+        // Get requester credit balance
+        uint256 balance = mapRequesterBalances[requester];
+        // Get requester actual subscription balance
+        uint256 subscriptionBalance = IERC1155(subscriptionNFT).balanceOf(requester, subscriptionTokenId);
+
+        // Adjust requester balance with maxDeliveryRate credits
+        balance += maxDeliveryRate;
+
+        // Check the request delivery rate for a fixed price
+        if (subscriptionBalance < balance) {
+            revert InsufficientBalance(subscriptionBalance, balance);
+        }
```

Comment:
Note the difference - we figure out the actual subscription amount and how much the next request costs. The overall value must not be bigger than the balance in credits already accounted for.